File: mobile/src/features/push/services/notificationRoute.ts
// Maps an incoming push message to the in-app route the user should land on
// when they tap it (or when we render an in-app toast for a foreground push).
//
// Two payload shapes are supported:
//
// 1. Modern, structured (`data.kind` + `data.entity_id` + optional
//    `data.conversation_id`). Emitted by `public.dispatch_push` from
//    20260606150000_dispatch_push_payload.sql onwards. Lets the client route
//    deterministically without parsing server-rendered strings.
//
// 2. Legacy (`data.url`) — the SQL still emits this for backward compat with
//    older app builds and as a fallback for the dev stub.
//
// `resolveNotificationRoute` prefers (1) when `kind` is present and falls
// through to (2) otherwise. Returns `null` when no route can be derived.

export type PushDataLike = {
  kind?: string;
  entity_id?: string;
  conversation_id?: string;
  url?: string;
  [key: string]: unknown;
};

export function resolveNotificationRoute(data: PushDataLike | undefined | null): string | null {
  if (!data) return null;

  const kind = typeof data.kind === 'string' ? data.kind : undefined;
  const entityId = typeof data.entity_id === 'string' ? data.entity_id : undefined;
  const conversationId =
    typeof data.conversation_id === 'string' ? data.conversation_id : undefined;
  const legacyUrl = typeof data.url === 'string' ? data.url : undefined;

  if (kind) {
    switch (kind) {
      case 'intro_received':
        return entityId ? `/(app)/intros/${entityId}` : '/(app)/(tabs)/inbox';

      // Direct message kinds — entity_id is the *message* id, not the
      // conversation id. We can only route to the chat when SQL also passed
      // conversation_id; otherwise we fall back to the chat list.
      case 'message_received':
      case 'image_received':
      case 'voice_received':
        return conversationId
          ? `/(app)/chats/${conversationId}`
          : '/(app)/(tabs)/chats';

      case 'meeting_proposal':
      case 'meeting_confirmed':
        return conversationId
          ? `/(app)/chats/${conversationId}`
          : '/(app)/(tabs)/inbox';

      // Opportunity interest — entity_id is the opportunity_id; route to detail.
      case 'opportunity_interest':
        return entityId
          ? `/(app)/opportunities/${entityId}`
          : '/(app)/(tabs)/opportunities';

      // Unknown kind — fall through to legacy url if SQL also supplied one.
      default:
        break;
    }
  }

  return legacyUrl ?? null;
}

File: mobile_flutter/lib/core/push/notification_route.dart
import '../routing/routes.dart';

/// Translates an FCM `data` map (and optional legacy `payload`) into a
/// go_router path. Spec section 7.4 verbatim:
///
/// | kind                                            | route                          | fallback           |
/// |-------------------------------------------------|--------------------------------|--------------------|
/// | intro_received, intro_accepted                  | /intros/{entity_id}            | /inbox             |
/// | message_received, image_received, voice_received| /chats/{conversation_id}       | /chats             |
/// | meeting_proposal, meeting_confirmed             | /chats/{conversation_id}       | /inbox             |
/// | opportunity_interest                            | /opportunities/{entity_id}     | /opportunities     |
/// | (unknown)                                       | payload.url                    | /home              |
///
/// FCM stringifies non-string `data` values on the way to the device, but on
/// the way IN (e.g. from a unit test that hand-builds the map) we may see a
/// raw `int`. `.toString()` covers both cases without forcing every caller to
/// pre-stringify.
String resolvePushRoute(
  Map<String, dynamic> data,
  Map<String, dynamic>? payload,
) {
  final String? kind = (data['kind'] as Object?)?.toString();
  final String? entityId = (data['entity_id'] as Object?)?.toString();
  final String? conversationId =
      (data['conversation_id'] as Object?)?.toString();

  switch (kind) {
    case 'intro_received':
    case 'intro_accepted':
      return (entityId != null && entityId.isNotEmpty)
          ? Routes.intro(entityId)
          : Routes.inbox;

    case 'message_received':
    case 'image_received':
    case 'voice_received':
      return (conversationId != null && conversationId.isNotEmpty)
          ? Routes.chat(conversationId)
          : Routes.chats;

    case 'meeting_proposal':
    case 'meeting_confirmed':
      return (conversationId != null && conversationId.isNotEmpty)
          ? Routes.chat(conversationId)
          : Routes.inbox;

    case 'opportunity_interest':
      return (entityId != null && entityId.isNotEmpty)
          ? Routes.opportunity(entityId)
          : Routes.opportunities;

    default:
      final String? legacy = (payload?['url'] as Object?)?.toString();
      if (legacy != null && legacy.isNotEmpty) return legacy;
      return Routes.home;
  }
}
